mlir-clang/mlir-clang.cc

```diff
@@ -153,8 +153,8 @@ int main(int argc, char **argv) {
     }
     return 0;
   }
-  auto module =
-      mlir::ModuleOp::create(mlir::OpBuilder(&context).getUnknownLoc());
+  mlir::OwningOpRef<mlir::ModuleOp> module(
```

Comment:
For sake of understanding, why does this need to become a pointer?

Reply:
we did not destroy module op. By wrapping it with OwningOpRef it will automatically destroyed. Does this answer your question?

Reply:
It should be destroyed by the deconstruct or at the end of the program though, right?

Reply:
Yes, exactly. If we don't want to wrap the module with OwningOpRef we need to insert the following lines at the end of the program:
```
if (module)
  module->erase();
```
The point with this is that we need to take care of doing that at the end of the program and we may forget, as we did. OwningOpRef to me looks like the right approach but if it is too invasive at this point the two extra lines are fine.

Reply:
Why do you think that OwningOpRef is not the right way to go?

Reply:
so my understanding of C++ is suhc that if you define it as a stack variable, it would have its deconstructor run at the end of the program. So for example:
```
void f() {
  Module m;
  ...
  // deconstruct m is run before any return automatically
}
```

I thought that's what we were doing, but perhaps I'm misremembering.

Reply:
Oh okay I see what was happening, yeah this is the way to go.